src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/GCStaticsPreInitDataNode.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+
+using Internal.Text;
+using Internal.TypeSystem;
+
+using Debug = System.Diagnostics.Debug;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    /// <summary>
+    /// Contains all GC static fields for a particular EEType.
+    /// Fields that have preinitialized data are pointer reloc pointing to frozen objects.
+    /// Other fields are initialized with 0.
+    /// We simply memcpy these over the GC static EEType object.
+    /// </summary>
+    public class GCStaticsPreInitDataNode : ObjectNode, IExportableSymbolNode
+    {
+        private MetadataType _type;
+        private List<PreInitFieldInfo> _sortedPreInitFields;
```

Comment:
A scheme like this:

```csharp
    class TypePreinitializationInfo
    {
        public MetadataType Type { get; }
        public bool HasInitializedGCStaticData { get; }

        public void WriteGCStaticData(ref ObjectDataBuilder builder, NodeFactory factory)
        {
            // Foreach field, call Write(). Insert padding where needed.
        }

        // Future: public void WriteNonGCStaticData(ref ObjectDataBuilder builder, NodeFactory factory)
    }

    abstract class FieldPreinitializationInfo
    {
        public FieldDesc Field { get; }

        public abstract void Write(ref ObjectDataBuilder builder, NodeFactory factory);
    }

    class ArrayFieldPreinitializationInfo : FieldPreinitializationInfo
    {
        public override void Write(ref ObjectDataBuilder builder, NodeFactory factory)
        {
            builder.EmitPointerReloc(factory.SerializedFrozenArray(this));
        }
    }
```

These can be allocated in a separate class (`PreinitializationManager`?) that is accessible from `NodeFactory`.

This will put us into a perfect place for introducing a `StaticInitDataTransform` later.